services/can_logger/can_logger.py

```diff
@@ -59,93 +78,236 @@ def db_init():
+
+        # Open the file and read it
+        with open(path) as typefile:
+            networktype = typefile.read().strip()
+
+        # 280 is the type for CAN. 'Documentation' here:
+        # https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/if_arp.h#L56
+        if networktype.isdigit() and int(networktype) == 280:
+            print('\t', network, ' appears to be a CAN link')
+            can_interfaces.append(network)
+
+    if len(can_interfaces) <= 0:
+        print('FATAL: No CAN interfaces found')
+        sys.exit(-1)
+
+    print(len(can_interfaces), ' found: ', can_interfaces)
+
+    return can_interfaces
+
+# Return 'True' if the interface exists and 'False' if it does not.
+
+
+def check_can_interface(interface, avail_interfaces):
+
+    for i in avail_interfaces:
+
+        if(interface == i):
+            return True
+
+    return False
+
+
+def log_can(can_interface):
+
+    print(f'Logging {can_interface}')
+
+    rx_buff = []
+    wr_buff = []
+    socket_connected = False
+
+    # Initialize socket
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+
+    # Connect to socketcand
+
+    while(socket_connected is False):
+
+        try:
+
+            s.connect((host_ip, int(host_port)))
+
+        except(ConnectionRefusedError):
+
+            print('Could not connect to socketcand. Connection Refused. Retrying...')
+            time.sleep(10)
+            socket_connected = False
+
+        else:
+
+            print(
+                f'Successfully connected to socketcand at {host_ip}: {host_port}')
+            socket_connected = True
+
+    # Receive socketcand's response
+
+    s.recv(32)
+
+    # Connect to exposed CAN interface
+
+    s.sendall(b'< open ' + can_interface.encode('utf-8') + b' >')
+
+    s.recv(32)
+
+    # Set socket to 'rawmode' to receive every frame on the bus.
+    s.sendall(b'< rawmode >')
+
+    s.recv(32)
+
+    # Receive frames in a 128-byte buffer and then decode the bytes object.
+    # Then strip some characters to clean up the received frame and then split 
+    # the resulting string to get the timestamp, CAN ID and CAN frame.
+
+    while(True):
+
+        frame = s.recv(64)
+        frame = frame.decode("utf-8").strip("<>''").split(' ')
+
+        try:
+
+            (timestamp, can_bus, can_id, can_data) = (frame[3], can_interface,
+                                                      frame[2], frame[4])
+
+        except(IndexError):
+
+            print(f'No CAN frames on {can_interface}. Waiting...')
+            time.sleep(1)
+
+        else:
+
+            timestamp = datetime.fromtimestamp(float(timestamp)).isoformat()
+            rx_buff.append((timestamp, can_bus, can_id, can_data))
+
+        # When the receive buffer reaches 100 entries, copy data from receive
+        # buffer to write buffer, then write to database from write buffer and
+        # clear receive buffer to continue receiving data
+
+        if(len(rx_buff) == 100):
```

Comment:
```suggestion
        if(len(rx_buff) >= 100):
```

Very minor but if avoids issues if `rx_buff` somehow jumps above 100 elements. Just a habit I built got my HS programming instructor